README.md
Python Timer Lite

This relatively light script allows you to run a simple command line timer to log time against a certain task.

Properties file should be updated with a system where you have ssh access (you should use a key to authenticate) and the folder where the timesheet is saved. Ensure you enable remote_save if you want this option.

```
host=timer # name of host from your ssh/config
path=~/timer/ # folder where you want the timesheet csv saved
remote_save=true # whether you want to save the file remotely
```


CHANGELOG.md
12-09-2023 - version 0.04
=========
tinker.py implements the same code but with TKinter to provide a simpler interface and a summary page

19-05-2023 - version 0.03
========== 

- show more info about prior tasks


12-05-2023 - version 0.02
========== 

- fix screen flickers

10-05-2023 - version 0.01
========== 

- added support for ssh storage of the timesheet.csv this allows you to store the timesheet remotely and use script from any system as long as you have a remote system you can ssh to. Ensure that your ssh/config file has an entry that facilitates key access to make this smooth.
- added logging
- download is made again before upload when stopping the timer to ensure we are using the latest timesheet when recording time
- allow user to choose from the last 3 tasks






timer.py
import time
import csv
import math
import curses
import subprocess
import hashlib
import os
import logging

# Set up logging
logging.basicConfig(filename="timer.log", level=logging.INFO, format="%(asctime)s - %(levelname)s - %(message)s")

#...
def last_descriptions_from_csv(n=3):
    """Retrieve the last n descriptions from the CSV file, along with their dates and durations"""
    records = []
    with open("timesheet.csv", newline="") as f:
        reader = csv.reader(f)
        for row in reader:
            if len(row) >= 4:
                records.append(row[:4])  # Get the date, duration, and description
    return records[-n:]

def display_timesheet(stdscr):
    """Display all entries from the timesheet at the bottom of the screen"""
    try:
        with open("timesheet.csv", newline="") as f:
            reader = csv.reader(f)
            y, x = stdscr.getmaxyx()  # get the height and width of the screen
            for i, row in enumerate(reversed(list(reader))):
                # If i is equal to or greater than y, break the loop
                if i >= y:
                    break
                row_str = " | ".join(row)
                if len(row_str) > x:
                    row_str = row_str[:x-3] + "..."
                stdscr.addstr(y - i - 1, 0, row_str)  # use y - i - 1 instead of y - i
    except FileNotFoundError:
        pass


def check_git_pull():
    """Performs a git pull and checks if any changes were made."""
    result = subprocess.run(["git", "pull"], capture_output=True, text=True)
    logging.info(f"git check results: {result}")
    return "Already up to date." not in result.stdout

def parse_properties(filename):
    properties = {}
    with open(filename, "r") as f:
        for line in f:
            key, value = line.strip().split("=", 1)
            properties[key] = value
    return properties

def download_timesheet(properties):
    if not properties['remote_save'] == "true":
        return
    """Downloads timesheet.csv from the remote system using SCP."""
    logging.info(f"Downloading latest version of timesheet...")
    remote_path = f"{properties['host']}:{properties['path']}/timesheet.csv"
    subprocess.run(["scp", remote_path, "timesheet.csv"])

def upload_timesheet(properties):
    if not properties['remote_save'] == "true":
        return
    """Uploads timesheet.csv to the remote system using SCP."""
    logging.info(f"Uploading latest version of timesheet...")
    remote_path = f"{properties['host']}:{properties['path']}/timesheet.csv"
    subprocess.run(["scp", "timesheet.csv", remote_path])

def format_time(duration):
    """Format duration as hours, minutes, and seconds"""
    hours, remainder = divmod(duration, 3600)
    minutes, seconds = divmod(remainder, 60)
    return "{:02d}:{:02d}:{:02d}".format(int(hours), int(minutes), int(seconds))

def save_time(description, project, elapsed_time_seconds):
    """Save elapsed time to CSV file"""
    elapsed_time_hours = int(elapsed_time_seconds / 3600)
    elapsed_time_minutes = int((elapsed_time_seconds % 3600) / 60)
    elapsed_time_seconds = int(elapsed_time_seconds % 60)
    elapsed_time_str = "{:02d}:{:02d}:{:02d}".format(elapsed_time_hours, elapsed_time_minutes, elapsed_time_seconds)
    start_time_str = time.strftime("%Y-%m-%d %H:%M:%S", time.localtime(start_time))
    with open("timesheet.csv", "a", newline="") as f:
        writer = csv.writer(f)
        writer.writerow([start_time_str, elapsed_time_str, description, project])

def get_project_list():
    """Get list of projects from CSV file"""
    project_set = set()
    try:
        with open("timesheet.csv", newline="") as f:
            reader = csv.reader(f)
            for row in reader:
                if len(row) > 3:
                    project_set.add(row[3])
    except FileNotFoundError:
        pass

    return sorted(list(project_set))

def select_project(stdscr, project_list):
    """Select project from list"""
    project_list.append("Enter a new project")
    selected_index = 0
    while True:
        stdscr.clear()
        stdscr.addstr(0, 0, "Select a project:")
        for i, project in enumerate(project_list):
            if i == selected_index:
                stdscr.addstr(i+2, 0, "> {}".format(project))
            else:
                stdscr.addstr(i+2, 0, "  {}".format(project))
        stdscr.refresh()
        c = stdscr.getch()
        if c == curses.KEY_UP:
            selected_index = max(0, selected_index-1)
        elif c == curses.KEY_DOWN:
            selected_index = min(len(project_list)-1, selected_index+1)
        elif c == ord('\n'):
            if selected_index == len(project_list) - 1:  # "Enter a new project" is selected
                return None
            else:
                return project_list[selected_index]

def input_project(stdscr):
    stdscr.addstr(1, 0, "Enter a project:")
    stdscr.refresh()
    project = ""
    while True:
        c = stdscr.getch()
        if c == ord('\n'):
            break
        elif c == curses.KEY_BACKSPACE or c == 127:
            project = project[:-1]
        elif c != -1:  # Ignore timeout (-1)
            project += chr(c)
        stdscr.clear()
        stdscr.addstr(1, 0, "Enter a project: {}".format(project))
        stdscr.refresh()
    return project

def select_description(stdscr, records):
    """Select a description from the last 3 tasks or enter a new one"""
    selected_index = -1
    custom_description = ""

    # define the headers
    headers = ["Date", "Duration", "Description", "Project"]

    while True:
        stdscr.erase()
        stdscr.addstr(0, 0, "Select a description or enter a new one:")

        # draw the table headers
        for i, header in enumerate(headers):
            stdscr.addstr(2, i*25, header)

        for i, record in enumerate(records):
            date, duration, desc, project = record

            # Highlight selected row
            if i == selected_index:
                stdscr.attron(curses.A_REVERSE)

            stdscr.addstr(i+3, 0, f"{date:20}")
            stdscr.addstr(i+3, 25, f"{duration:20}")
            stdscr.addstr(i+3, 50, f"{desc[:20]:20}")
            stdscr.addstr(i+3, 75, f"{project[:20]:20}")

            if i == selected_index:
                stdscr.attroff(curses.A_REVERSE)

        stdscr.addstr(len(records)+3, 0, "Custom: {}".format(custom_description))
        stdscr.refresh()

        c = stdscr.getch()
        if c == curses.KEY_UP:
            selected_index = max(0, selected_index - 1)
        elif c == curses.KEY_DOWN:
            selected_index = min(len(records) - 1, selected_index + 1)
        elif c == ord('\n'):
            return records[selected_index][2] if selected_index >= 0 else custom_description
        elif c == curses.KEY_BACKSPACE or c == 127:
            custom_description = custom_description[:-1]
        elif c != -1:  # Ignore timeout (-1)
            custom_description += chr(c)


def main(stdscr):
    """Main function"""
     # Read properties from the timer.properties file
    properties = parse_properties("timer.properties")

    # Perform a git pull to ensure the latest version of the code
    if check_git_pull():
        stdscr.clear()
        stdscr.addstr(0, 0, "A new version of the script has been downloaded.")
        stdscr.addstr(1, 0, "Please restart the script.")
        stdscr.refresh()
        while True:
            c = stdscr.getch()
            if c == ord('q'):
                return

    # Download timesheet.csv from the remote system
    download_timesheet(properties)


    global start_time
    curses.curs_set(1)
    curses.init_pair(1, curses.COLOR_WHITE, curses.COLOR_BLACK)
    curses.init_pair(2, curses.COLOR_BLACK, curses.COLOR_WHITE)
    stdscr.bkgd(curses.color_pair(2))
#    stdscr.timeout(1000)  # Set timeout to 1000 ms (1 second)

# Display all entries from the timesheet.csv
    display_timesheet(stdscr)

    # Select or enter a description
    last_descriptions = last_descriptions_from_csv(10)
    if last_descriptions:
        stdscr.addstr(0, 0, "Select a description:")
        stdscr.refresh()
        description = select_description(stdscr, last_descriptions)
        if description:
            # Extract the actual description from the selected option
            description = description.split(" - ", 2)[-1]
    else:
        stdscr.addstr(0, 0, "Enter a description:")
        stdscr.refresh()
        description = ""
        while True:
            c = stdscr.getch()
            if c == curses.KEY_ENTER or c == 10 or c == 13:
                break
            elif c == curses.KEY_BACKSPACE:
                description = description[:-1]
            else:
                description += chr(c)
            stdscr.erase()  # Use erase instead of clear
            stdscr.addstr(0, 0, "Enter a description: {}".format(description))
            stdscr.refresh()


    # Select or enter a project
    project_list = get_project_list()
    if project_list:
        stdscr.addstr(1, 0, "Select a project:")
        stdscr.refresh()
        project = select_project(stdscr, project_list)
        if project is None:  # "Enter a new project" was selected
            project = input_project(stdscr)
    else:
        project = input_project(stdscr)

    # Start the timer
    stdscr.clear()
    stdscr.addstr(0, 0, "Press 's' to start the timer...")
    stdscr.refresh()
    stdscr.timeout(500) # Set timeout to 1000 ms (1 second)
    
    while True:
        c = stdscr.getch()
        if c == ord('s'):
            stdscr.clear()
            start_time = time.time()
            start_time_str = time.strftime("%Y-%m-%d %H:%M:%S", time.localtime(start_time))
            stdscr.addstr(0, 0, "Timer started at {}".format(start_time_str))
            stdscr.addstr(1, 0, "Description: {}".format(description))
            stdscr.addstr(2, 0, "Project: {}".format(project))
            elapsed_time = time.time() - start_time
            stdscr.addstr(3, 0, "Elapsed time: {}".format(format_time(elapsed_time)), curses.color_pair(1))
            stdscr.refresh()
            break

    # Monitor and display the elapsed time
    while True:
        c = stdscr.getch()
        if c == ord('q'):
            break
        elif c == ord('p'):
            stdscr.addstr(0, 0, "Timer stopped at {}".format(time.strftime("%Y-%m-%d %H:%M:%S")))
            stdscr.refresh()
            elapsed_time = time.time() - start_time
            
            # Download again in case another system has updated this file since we last checked
            download_timesheet(properties)
            
            save_time(description, project, elapsed_time)
            
            # Upload timesheet.csv to the remote system
            upload_timesheet(properties)
            
            break
        current_time = time.time()
        elapsed_time = current_time - start_time
        stdscr.addstr(3, 0, "Elapsed time: {}".format(format_time(elapsed_time)), curses.color_pair(1))
        stdscr.refresh()
        time.sleep(1)

# Start curses
curses.wrapper(main)


tinker.py
from datetime import datetime, timedelta

import time
import csv
import subprocess
import logging
from tkinter import Tk, Label, Button, Entry, Listbox, StringVar, Toplevel, Text, BOTH, YES, NONE, DISABLED, Scrollbar, RIGHT, Frame, Canvas, Y

from tkinter import ttk
from tkinter.ttk import Combobox
from datetime import datetime
from collections import defaultdict

def format_timedelta(td):
    total_seconds = int(td.total_seconds())
    hours, remainder = divmod(total_seconds, 3600)
    minutes, seconds = divmod(remainder, 60)
    return f"{hours} hours {minutes} minutes {seconds} seconds"


def check_git_pull():
    """Performs a git pull and checks if any changes were made."""
    result = subprocess.run(["git", "pull"], capture_output=True, text=True)
    logging.info(f"git check results: {result}")
    return "Already up to date." not in result.stdout

def save_time(description, project, elapsed_time_seconds,start_time_str):
        """Save elapsed time to CSV file"""
        elapsed_time_hours = int(elapsed_time_seconds / 3600)
        elapsed_time_minutes = int((elapsed_time_seconds % 3600) / 60)
        elapsed_time_seconds = int(elapsed_time_seconds % 60)
        elapsed_time_str = "{:02d}:{:02d}:{:02d}".format(elapsed_time_hours, elapsed_time_minutes, elapsed_time_seconds)
        
        with open("timesheet.csv", "a", newline="") as f:
            writer = csv.writer(f)
            writer.writerow([start_time_str, elapsed_time_str, description, project])

def download_timesheet(properties):
    if not properties['remote_save'] == "true":
        return
    """Downloads timesheet.csv from the remote system using SCP."""
    logging.info(f"Downloading latest version of timesheet...")
    remote_path = f"{properties['host']}:{properties['path']}/timesheet.csv"
    subprocess.run(["scp", remote_path, "timesheet.csv"])


def upload_timesheet(properties):
    if not properties['remote_save'] == "true":
        return
    """Uploads timesheet.csv to the remote system using SCP."""
    logging.info(f"Uploading latest version of timesheet...")
    remote_path = f"{properties['host']}:{properties['path']}/timesheet.csv"
    subprocess.run(["scp", "timesheet.csv", remote_path])
    
def parse_properties(filename):
    properties = {}
    with open(filename, "r") as f:
        for line in f:
            key, value = line.strip().split("=", 1)
            properties[key] = value
    return properties

# Set up logging
logging.basicConfig(filename="timer.log", level=logging.INFO, format="%(asctime)s - %(levelname)s - %(message)s")


class TimesheetApp:
    def __init__(self, master):
        self.master = master
        self.master.title("Timesheet App")
        
        self.start_time = None
        self.project_var = StringVar()
        
        # Initialize labels for description and project titles
        self.description_label = Label(master, text="")
        self.project_label = Label(master, text="")
        self.timer_label = Label(master, text="", font=("Helvetica", 35))
        
        # Get unique descriptions and projects and the last description and project from timesheet.csv
        self.unique_descriptions, self.unique_projects, last_description, last_project = self.get_unique_data()
        
        self.description_combobox = Combobox(master, values=self.unique_descriptions)
        self.description_combobox.pack()
        if last_description:
            self.description_combobox.set(last_description)
        else:
            self.description_combobox.set("Select or type a description")
        
        self.project_combobox = Combobox(master, values=self.unique_projects)
        self.project_combobox.pack()
        if last_project:
            self.project_combobox.set(last_project)
        else:
            self.project_combobox.set("Select or type a project")
        
        self.start_button = Button(master, text="Start Timer", command=self.start_timer)
        self.stop_button = Button(master, text="Stop Timer", command=self.stop_timer)
        self.save_button = Button(master, text="Save Entry", command=self.save_entry)
        self.start_button.pack()
        
        self.details_button = Button(master, text="Summary", command=self.show_details_window)
        self.details_button.pack()
        
        self.update_clock()

    def get_descriptions_with_durations(self):
        data = []
        monthly_data = {}

        try:
            with open('timesheet.csv', 'r') as f:
                reader = csv.reader(f)
                for row in reader:
                    if row:
                        start_date_time, elapsed_time, description, project = row
                        start_dt = datetime.strptime(start_date_time, '%Y-%m-%d %H:%M:%S')
                        elapsed_h, elapsed_m, elapsed_s = map(int, elapsed_time.split(':'))
                        elapsed_td = timedelta(hours=elapsed_h, minutes=elapsed_m, seconds=elapsed_s)

                        month_key = start_dt.strftime('%Y-%m')
                        if month_key not in monthly_data:
                            monthly_data[month_key] = {'entries': [], 'total_duration': timedelta()}

                        monthly_data[month_key]['entries'].append({
                            'Date': start_dt.strftime('%Y-%m-%d'), 
                            'Description': description, 
                            'Duration': elapsed_td
                        })
                        monthly_data[month_key]['total_duration'] += elapsed_td

            for month_key, month_data in monthly_data.items():
                month_name = datetime.strptime(month_key, '%Y-%m').strftime('%B %Y')  # Convert month key to month name
                data.append({'Date': f'Month: {month_name}', 'Description': '', 'Duration': ''})
                data.extend(month_data['entries'])
                data.append({
                    'Date': 'Monthly Total', 
                    'Description': '', 
                    'Duration': format_timedelta(month_data['total_duration'])  # Use the new format
                })

        except FileNotFoundError:
            pass

        return data



    def show_details_window(self):
        new_window = Toplevel(self.master)
        new_window.title("Details Window")
 
        # Set initial size
        new_window.geometry("800x600")  # Adjust width (800) and height (600) as per your requirement
    

        canvas = Canvas(new_window)
        scrollbar = Scrollbar(new_window, orient="vertical", command=canvas.yview)
        scrollable_frame = Frame(canvas)

        canvas.configure(yscrollcommand=scrollbar.set)

        # Bind the mouse wheel event to the canvas scroll
        #canvas.bind("<MouseWheel>", lambda event: canvas.yview_scroll(-1*(event.delta//120), "units"))
        new_window.bind("<MouseWheel>", lambda event: canvas.yview_scroll(-1*(event.delta//120), "units"))

        
        scrollable_frame.bind(
            "<Configure>",
            lambda e: canvas.configure(
                scrollregion=canvas.bbox("all")
            )
        )

        canvas.create_window((0, 0), window=scrollable_frame, anchor="nw")
        canvas.configure(yscrollcommand=scrollbar.set)

        descriptions_with_durations = self.get_descriptions_with_durations()
        bg_color1 = "#703224"  # A light gray color

        for index, row_data in enumerate(descriptions_with_durations):
            if "Month:" in row_data['Date']:
                Label(scrollable_frame, text=row_data['Date'], width=70, bg=bg_color1, anchor='w').grid(row=index, column=0, columnspan=4)
            elif "Monthly Total" in row_data['Date']:
                #Label(scrollable_frame, text=row_data['Date'], width=10, anchor='w').grid(row=index, column=0)
                #Label(scrollable_frame, text=row_data['Description'], width=50, anchor='w').grid(row=index, column=1)
                Label(scrollable_frame, text=row_data['Duration'], width=70,bg=bg_color1, anchor='center').grid(row=index, column=0, columnspan=4)  # Center align the duration
            else:
                Label(scrollable_frame, text=row_data['Date'], width=10).grid(row=index, column=0)
                Label(scrollable_frame, text=row_data['Description'], width=50).grid(row=index, column=1)
                Label(scrollable_frame, text=row_data['Duration'], width=10).grid(row=index, column=2)

        canvas.pack(side="left", fill="both", expand=True)
        scrollbar.pack(side="right", fill="y")

    def get_unique_data(self):
        descriptions = set()
        projects = set()
        last_description = None
        last_project = None

        try:
            with open('timesheet.csv', 'r') as f:
                reader = csv.reader(f)
                for row in reader:
                    if row:
                       # print(row)
                        descriptions.add(row[2])  # Assuming description is in the first column
                        projects.add(row[3])  # Assuming project is in the second column
                        last_description = row[2]
                        last_project = row[3]
        except FileNotFoundError:
            pass  # File not found, return empty sets and None values

        return list(descriptions), list(projects), last_description, last_project

        
    def start_timer(self):
        # Get the selected description and project
        self.description = self.description_combobox.get()
        self.project = self.project_combobox.get()

        # Hide the input fields and display the timer and titles
        self.description_combobox.pack_forget()
        self.project_combobox.pack_forget()
        self.start_button.pack_forget()
        
        self.project_label.config(text=f"{self.project}")
        self.project_label.pack()

        self.description_label.config(text=f"{self.description}")
        self.description_label.pack()

        self.start_time = time.time()
        self.timer_label.pack()
        # Display the stop button when the timer starts
        self.stop_button.pack()
        self.update_clock()
    
    def stop_timer(self):
        self.start_time_str  = time.strftime("%Y-%m-%d %H:%M:%S", time.localtime(self.start_time))
        self.elapsed_time = time.time() - self.start_time

        # Hide the timer label and stop button, and show the input fields again
        self.stop_button.pack_forget()        
        self.save_button.pack()
        # Reset the start time to stop the timer
        self.start_time = None
        
    def save_entry(self):
        self.save_button.pack_forget()
        # Download again in case another system has updated this file since we last checked
        download_timesheet(properties)
        save_time(self.description,self.project, self.elapsed_time,self.start_time_str)
        # Upload timesheet.csv to the remote system
        upload_timesheet(properties)
        logging.info('done uploading new time entry, shutting down timer')
        self.master.destroy()

    def update_clock(self):
        if self.start_time:
            elapsed_time = time.time() - self.start_time
            mins, sec = divmod(elapsed_time, 60)
            hours, mins = divmod(mins, 60)
            self.timer_label.config(text="Elapsed time: {:02d}:{:02d}:{:02d}".format(int(hours), int(mins), int(sec)))
            
            # Update the timer label every second
            self.master.after(1000, self.update_clock)
        

    def format_time(self, duration):
        """Format duration as hours, minutes, and seconds"""
        hours, remainder = divmod(duration, 3600)
        minutes, seconds = divmod(remainder, 60)
        return "{:02d}:{:02d}:{:02d}".format(int(hours), int(minutes), int(seconds))

# Run the Tkinter app
if __name__ == "__main__":
    root = Tk()
    # Read properties from the timer.properties file
    properties = parse_properties("timer.properties")
    # Download timesheet.csv from the remote system
    download_timesheet(properties)
    app = TimesheetApp(root)
    root.mainloop()

